### PR TITLE
Updated Rewrite delimcpy to use memchr and Copy, not per-byte #18056

### PR DESCRIPTION
**Pull request by khwilliamson**
I mistakenly closed #18056. This is an updated version of that

### CODE REVIEW

File: ext/XS-APItest/APItest.xs [hunk 1]
```diff
@@ -6795,6 +6795,31 @@ test_toTITLE_utf8(SV * p, int type)
     OUTPUT:
         RETVAL
 
+AV *
+test_delimcpy(SV * from_sv, STRLEN trunc_from, char delim, STRLEN to_len, STRLEN trunc_to, char poison = '?')
```
Comment (tonycoz) [inline ext/XS-APItest/APItest.xs:6799]: It's not a big deal for test code like this (unless we valgrind or ASAN), but the default-for-backward-compatability AV* typemap leaks.

This should include 
```
AV*	T_AVREF_REFCOUNT_FIXED
```

in the typemap to get the non-leaky version.

**Comment by tonycoz** [comment]
The change in failure return is dangerous.

1) simply making a pointer that points outside of the buffer (before the first element, beyond the point just after the last element) produces undefined behaviour.

2) on 32-bit platforms, if the buffer is allocated in the top 2G of the address space this will cause the pointer to wrap around, and code that checks if the scan was successful will pass even though it shouldn't (one reason for 1)

3) code that checks specifically for the end pointer, like this from toke.c will be broken:

```
	s = delimcpy(d, e, s, PL_bufend, term, &len);
	if (s == PL_bufend)
	    Perl_croak(aTHX_ "Unterminated delimiter for here document");
```

**Comment by khwilliamson** [comment]
I think you misunderstand what the changes are.  I did not change the return value of the function, so in your example `s` is unchanged from prior to the commit.  The failure return is instead in  the `&len` component.   I've changed that to be a large number.  It's the same  paradigm we commonly use of  `*retlen = (STRLEN)-1`.

I should have mentioned in the commit message that in blead, this function does not indicate failure in any way.  And that is dangerous.  How do you add a failure return to an API function, without changing existing behavior when there isn't an error?  This is what I came up with.  Instead of returning -1, though, I noticed that core code did calculations on `len` which could easily have pushed it over to a small positive number., so I chose IV_MAX instead.

But your confusion  means the doc wording should be improved.  What do you suggest?

**Comment by tonycoz** [comment]
> But your confusion means the doc wording should be improved. What do you suggest?

I think it was the commit message that misled me:

```
It also changes the error return which happens when the Copy won't fit
in the allowed space.  Previously the error return (in *retlen) was
to just beyond the end of the destination buffer.
```

Sorry.

That said, if this API was widely used I'd still think it was a risky change, code that passed in a short buffer (that had more space after the buffer supplied to delimcpy for local use) that modifies the returned length (which is in an I32) might push it into a negative number (or produce undefined behaviour)

But it isn't widely used, I only see a few uses in core and none on CPAN, so you can ignore that if you wish.

I suspect there's a change in behaviour when delim is `'\\'`, but none of the callers in core use that as far as I can see.

File: ext/XS-APItest/APItest.xs (hunk 1 above)
Comment (khwilliamson) [reply ext/XS-APItest/APItest.xs:6799]: grepping, I didn't find an example of how to do this.  There are a bunch of functions in this .xs that return AV's

**Comment by khwilliamson** [comment]
> I suspect there's a change in behaviour when delim is '\\', but none of the callers in core use that as far as I can see.

Indeed there is a change.  You apparently are the first person to consider the possibility of the delimiter being a backslash.  The blead code just doesn't work for that, ignoring any backslashes in the buffer and returning the full length.

File: ext/XS-APItest/APItest.xs (hunk 1 above)
Comment (tonycoz) [reply ext/XS-APItest/APItest.xs:6799]: The fix only needs the line above to be added to the top section of `ext/XS-APItest/typemap`.

Unless they have an extra sv_2mortal() or used the fixed typemap entry they'll leak the AV, see T_AVREF in perlxstypemap.

**Pull request closed without merging**